### PR TITLE
Add --benchmark_output=<filename> option.

### PR DESCRIPTION
**Pull request by EricWF**
This patch adds support for writing benchmark results to a file. The file is specified using `--benchmark_output=<filename>`. The output format used is specified by `--benchmark_output_format={json:csv:console}`.

This patch is intended to help support additional tooling which needs to be able to access the data in a less human friendly manner than the colored console reporter.

Specifying `--benchmark_output` does not change the console output in any way. This is useful when you want to both (A) watch the output and (B) save it for later.

In order to support the `console` output format when writing to files I had to clean up how `ConsoleReporter` handles color printing.

I have no idea how to test this change since the tests can't (shouldn't?) be creating files. Suggestions welcome.

### CODE REVIEW

**Comment by EricWF** [comment]
I honestly have no clue what's causing AppVeyorBot to fail :-S

**Comment by dominichamon** [comment]
appveyor failure is some weirdness with ctest. i'm working on a fix (slowly) but not making much progress.

File: README.md [hunk 1]
```diff
@@ -493,6 +493,12 @@ name,iterations,real_time,cpu_time,bytes_per_second,items_per_second,label
 "BM_SetInsert/1024/10",106365,17238.4,8421.53,4.74973e+06,1.18743e+06,
 ```
 
+## Output Files
+The library supports writing the output of the benchmark to a file specified
+by `--benchmark_output=<filename>`. The format of the output can be specified
+using `--benchmark_output_format={json|console|csv}`. Specifying
+`--benchmark_output` does not suppress the console output.
```
Comment (dominichamon) [inline README.md]: why not just invoke and pipe to tee then? i'm not sure that this library should be writing files at all, given that every OS has a way of redirecting output to a file.

Comment (EricWF) [inline README.md]: There's a big difference between output for humans and output for tooling. Simply put I can't grok JSON and tooling can't grok colored console output, so I don't want to choose between the two. I want the both.

I agree the library shouldn't write to files in general, but `-o output` is a pretty universal feature.

Comment (dominichamon) [inline README.md]: how often do you think someone would want to push the output to tooling and have it human readable? I'd expect any tooling requirement to override humans reading the output. ie, i think if you're going to push the output to a tool that tool should take care of the human readable bit too. whether that's a summary, or reformatting the raw results.

ie, i'd expect `$ benchmark_my_stuff --benchmark_format=json | analyse_my_bench`

to provide me with human readable analysis results. if i need the output for debugging, then `tee` or some equivalent would at least give me something to look at after the fact.

```
$ benchmark_my_stuff --benchmark_output=out.json --benchmark_output_format=json --benchmark_format=csv
```

it's very easy to get wrong. i mean, output is also to stdout, so output_format vs format is subtle.

this may just be bikeshedding the arguments, i guess. i'm wary of redundancy and confusion.

**Comment by EricWF** [comment]
@dominichamon Do you have an issue with the coverage drop here? I don't want to write tests that create output files since it's messy to clean them up.

**Comment by dominichamon** [comment]
no, i have no issue with the coverage drop or trying to test code that writes files. i'm still wary of the command line flag confusion.

File: README.md (hunk 1 above)
Comment (EricWF) [inline README.md]: > how often do you think someone would want to push the output to tooling and have it human readable? 

I do it all the time. I want to see the benchmark results as the come in, and I also want to do something with the results later.

> , i think if you're going to push the output to a tool that tool should take care of the human readable bit too. 

I think that tool should take care of the human readable input for the information it presents, I don't expect the tool to repeat the benchmark output. 

> ie, i'd expect $ benchmark_my_stuff --benchmark_format=json | analyse_my_bench
> to provide me with human readable analysis results. if i need the output for debugging, then > tee or some equivalent would at least give me something to look at after the fact.

I wouldn't want to require "analyse_my_bench" to know how to pretty print basic benchmark output better than the library can already do.

> ```$ benchmark_my_stuff --benchmark_output=out.json --benchmark_output_format=json --benchmark_format=csv```
> it's very easy to get wrong. i mean, output is also to stdout, so output_format vs format is subtle.
> i'm wary of redundancy and confusion.

I'm sympathetic to that. A better way would be nice, or better names. However I don't think offering two output methods is too confusing. This behavior mirrors how LLVM's test runner LIT works. For example `$ lit -o out.json test.cpp` will give human readable output to stdout, and will print a copy of the JSON results to out.json.

In summary I agree with the general concern over confusion, but I think the feature is still useful to have.

Comment (dominichamon) [inline README.md]: ok, i'll buy that. bikeshedding then:

--benchmark_out: the file to which to write results
--benchmark_out_format: the format in which to write the results to --benchmark_out, if set
--benchmark_format: the format in which to write the results to stdout

?

File: src/benchmark.cc [hunk 2]
```diff
@@ -1025,10 +1040,17 @@ std::unique_ptr<BenchmarkReporter> GetDefaultReporter() {
 } // end namespace internal
 
 size_t RunSpecifiedBenchmarks() {
```
Comment (dominichamon) [inline src/benchmark.cc]: why overload instead of using default arguments?

File: include/benchmark/reporter.h [hunk 3]
```diff
@@ -157,13 +157,18 @@ class BenchmarkReporter {
 // default reporter used by RunSpecifiedBenchmarks().
 class ConsoleReporter : public BenchmarkReporter {
  public:
+  explicit ConsoleReporter(bool color_output = true)
```
Comment (dominichamon) [inline include/benchmark/reporter.h]: one rather nice trick for avoiding boolean parameters (which can be confusing at the callsite) is to define an enum with two values.

File: src/benchmark.cc (hunk 2 above)
Comment (EricWF) [inline src/benchmark.cc]: Good question: It's to avoid the "zero as null pointer constant" warning GCC emits in C++98 mode. 
Optimally we would use "= nullptr" as the default argument, but that's C++11 only, and using either "NULL" or "0" as the default value will trigger the warning.

I would prefer to use default arguments though.

File: README.md (hunk 1 above)
Comment (EricWF) [inline README.md]: Sounds good to me. Changes incoming.

File: include/benchmark/reporter.h (hunk 3 above)
Comment (EricWF) [inline include/benchmark/reporter.h]: Ack. Changed in the newest revision.

**Comment by EricWF** [comment]
@googlebot confirmed.

File: include/benchmark/benchmark_api.h [hunk 4]
```diff
@@ -489,9 +494,9 @@ class Benchmark {
   // REQUIRES: The function passed to the constructor must accept an arg1.
   Benchmark* Range(int start, int limit);
 
-  // Run this benchmark once for every value in the range [start..limit]
+  // Run this benchmark once for all values in the range [start..limit] with specific step
```
Comment (dominichamon) [inline include/benchmark/benchmark_api.h]: why is this part of this PR?

**Comment by dominichamon** [comment]
LGTM

File: include/benchmark/benchmark_api.h (hunk 4 above)
Comment (EricWF) [inline include/benchmark/benchmark_api.h]: It's just git messing up history. This change is already in master, but I merged and caught the commit. I'll commit this patch without history.

**Comment by EricWF** [comment]
Sounds good. I'm going to commit directly to involve getting the history.
